source/multigrid/mg_transfer_internal.cc

```diff
@@ -564,6 +564,20 @@ namespace internal
                     continue;
                   cell->child(c)->get_mg_dof_indices(local_dof_indices);
 
+                  // Replace any indices of identity constrained dofs with the
+                  // indices of the dofs in which they are constrained
+                  if (mg_constrained_dofs != nullptr)
+                    for (unsigned int i=0; i<local_dof_indices.size(); ++i)
+                      if (mg_constrained_dofs->get_level_constraint_matrix(level).
+                          is_identity_constrained(local_dof_indices[i]))
+                        {
+                          Assert(mg_constrained_dofs->get_level_constraint_matrix(level).
+                                 get_constraint_entries(local_dof_indices[i])->size() == 1,
+                                 ExcInternalError());
+                          local_dof_indices[i] = mg_constrained_dofs->get_level_constraint_matrix(level).
+                                                 get_constraint_entries(local_dof_indices[i])->front().first;
+                        }
+
```

Comment:
The `MGTransferPrebuilt` result is still wrong (but we can't fix it), but the matrix-free result should be "correct" if we also add them. Did you test on an actual multigrid example with this commented in/out and check if there is any difference? Use somewhat higher polynomial degrees, say 3-5, when you check this.